### PR TITLE
Label Templates: add IP addresses to the Network variables

### PR DESCRIPTION
**Pull request by fgiudici**
Add the IPv4 and IPv6 addresses to the NICs template variables.
In particular:
* **${Network/NICs/_<number>_/IPv4Addresses/_<number>_}**
* **${Network/NICs/_<number>_/IPv6Addresses/_<number>_}**
* **${Network/NICs/_<number>_/IPv4Address}**   _# first address from IPv4Addresses_
* **${Network/NICs/_<number>_/IPv6Address}**   _# first address from IPv6Addresses_
* **${Network/NICs/_<ifaceName>_/IPv4Addresses/_<number>_}**
* **${Network/NICs/_<ifaceName>_/IPv6Addresses/_<number>_}**
* **${Network/NICs/_<ifaceName>_/IPv4Address}**
* **${Network/NICs/_<ifaceName>_/IPv6Address}**

Example:
```
apiVersion: elemental.cattle.io/v1beta1
kind: MachineRegistration
[...]
  machineInventoryLabels:
    element: fire
    eth1-IPv4: ${Network/NICs/eth1/IPv4Address}               # 10.100.103.144
    eth1-IPv6: ${Network/NICs/eth1/IPv6Address}               # fe80-922a-e5ea-fc77-e503 
    net0-name: ${Network/NICs/0/Name}                              # eth0
    net0-ipv4: ${Network/NICs/0/IPv4Address}                     # 172.21.100.153
    eth0-IPv4-0: ${Network/NICs/eth0/IPv4Addresses/0}    # 10.100.103.144
    eth0-IPv4-1: ${Network/NICs/eth0/IPv4Addresses/1}
    eth0-IPv6-0: ${Network/NICs/eth0/IPv6Addresses/0}    # fe80-922a-e5ea-fc77-e503 
    eth0-IPv6-1: ${Network/NICs/eth0/IPv6Addresses/1}
```
Part of #855

### CODE REVIEW

**Review by kkaempf — CHANGES_REQUESTED** [review]
Please add respective tests

File: pkg/util/net.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+	return getNMActivatedDevices(NewNetController())
+}
+
+func getNMActivatedDevices(nc NetController) (nmDevs []string) {
+	var buf *bytes.Buffer
+	var err error
+	nmDevs = []string{}
+
+	if buf, err = nc.GetNMDeviceState(); err != nil {
+		return
```
Comment (anmazzotti) [inline pkg/util/net.go:120]: Would be nice to print the error so that it does not fail silently, for example if `nmcli` is missing/broken.

**Review by anmazzotti — APPROVED** [review]
Looks good to me, just a few nitpicks.

However I'm worried about the implicit behavior of showing IPv4 *before* IPv6. Why not both? 

My proposal is to explicitly implement `*.ipv4` and `*.ipv6` annotations.
Makes it also easier to parse the value, so that you don't have to implement a `isIPv4?` filter for example.

Last but not least, this PR enables this behavior by default, I wonder if this is the desired behavior for everyone or maybe this is worth a new toggle, like a new flag in the MachineRegistration to enable or not reporting of IPs.
If we have the toggle I'd still enable it by default, as this is very nice to have.

**Review by anmazzotti — CHANGES_REQUESTED** [review]
Sorry I didn't mean to approve it, pressed the wrong button. :(

File: pkg/util/util.go [hunk 2]
```diff
@@ -79,7 +79,7 @@ func JSONObjectToYamlBytes(object map[string]runtime.RawExtension) ([]byte, erro
 	// This creates a parent "root" key to facilitate parsing the schemaless map
 	mapSlice := yaml.JSONObjectToYAMLObject(map[string]interface{}{"root": object})
 	if len(mapSlice) <= 0 {
-		return nil, errors.New("Could not convert json object to yaml")
```
[the remark this answers is not in the document]
Comment (fgiudici) [reply pkg/util/util.go:82]: It is just removing capitalized "Could" as hinted by go linter checks ("error message should not be capitalized"). Would you prefer to keep it capitalized?

**Comment by fgiudici** [comment]
> However I'm worried about the implicit behavior of showing IPv4 _before_ IPv6. Why not both?
> 
> My proposal is to explicitly implement `*.ipv4` and `*.ipv6` annotations. Makes it also easier to parse the value, so that you don't have to implement a `isIPv4?` filter for example.

Well, have to confess I was mainly looking at the final goal to have an IP address for k3s/RKE2 internal/external address values, and so ended up looking just for one IP address.. but good point, doesn't makes much sense to do this kind of filtering at the annotation level, would make more sense to have ipv4 and ipv6 👍🏼

File: pkg/util/net.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+		if ipNet, ok := addr.(*net.IPNet); ok {
+			res = append(res, *ipNet)
+		}
+	}
+	return
+}
+
+func (un *unixNet) GetNMDeviceState() (buf *bytes.Buffer, err error) {
+	buf = &bytes.Buffer{}
+	cmd := exec.Command("nmcli", "-g", "DEVICE,STATE", "device")
```
[the remark this answers is not in the document]
Comment (fgiudici) [reply pkg/util/net.go:64]: Yeah, we have a hard requirement on NetworkManager not explicitly  tracked in different places, in particular in our base images...
In the operator we need it for declarative networking, as all the configurators feed NetworkManager via its config dir.
Anyway, I have just changed my mind: putting default annotations with interfaces/ip addresses for each network device  doesn't look great. We should just support template labels to allow retrieving them on demand.
Going to drop this part (which is also quite hackish honestly).

**Comment by fgiudici** [comment]
Changed PR to just expose IP addresses as Network variables (on demand).
No more new default annotations.
We now expose IPv4 and IPv6 addresses, updated the PR description accordingly.

**Review by anmazzotti — APPROVED** [review]
Looks good, thank you for the changes!

**Pull request merged**